File: quick_sort_ffi.c
#include <lua.h>
#include <lauxlib.h>

static void swap(int* arr, size_t i, size_t j)
{
    int buf = arr[i];
    arr[i] = arr[j];
    arr[j] = buf;
}

static size_t partition(int* arr, int left, int right)
{
    int pivot_value = arr[right];
    size_t pivot = left - 1;

    for (int i = left; i < right; i++)
        if (arr[i] <= pivot_value)
            pm_swap(arr, i, ++pivot);

    pm_swap(arr, right, ++pivot);

    return pivot;
}

void tnt_quick_sort(int* arr, size_t left, size_t right)
{
    if (left >= 0 && left < right) {
        size_t pivot = partition(arr, left, right);
        qsort(arr, left, pivot - 1);
        qsort(arr, pivot + 1, right);
    }
}

File: quick_sort_ffi.lua
#!/usr/bin/env tarantool

local clock = require("clock")
local ffi = require("ffi")
local log = require("log")

math.randomseed(os.time())

local array = { }
local count = 10 * 1000 * 1000
for i = 1, count do
    array[i] = math.random(1000000)
end

ffi.cdef([[
    void tnt_quick_sort(int* arr, size_t left, size_t right);
]])

local package = package.search("mylib")
local mylib = ffi.load(package)

local start = tonumber(clock.proc64())
local arg = ffi.new(("int[%d]"):format(count), array)
mylib.tnt_quick_sort(arg, 0, count - 1)
local stop = tonumber(clock.proc64())

log.info("Time: %.2f sec", (stop - start) / 1000 / 1000 / 1000)
